### PR TITLE
Add link to agent a stage is running on to stage summary

### LINKED ISSUE #426: Show agent on pipeline console stage summary
### What feature do you want to see added?

On the pipeline console page, when a stage is selected on the left, show the agent the stage is running on in the summary at the top.
image

We have a pipeline that use a couple of labels to select nodes, one of those labels has 10 or so agents. When debugging a build failure, is common to ssh into the node and investigate. But it takes quite a few clicks to find which agent the stage is/was run on.

The workspace directory would be very useful too.

That info might take a bit too much vertical space, and there is plenty on horizontal space in than area, so perhaps it could go there.

### Upstream changes

_No response_

### Are you interested in contributing this feature?

_No response_

### PR DESCRIPTION
**Pull request by Hammie**
Adds a line to the console view stage summary, showing which agent the stage is running on. 

Fixes #426 

### Testing done

Ran the plugin on a Jenkins instance with multi-stage pipelines, and so far everything seems to work.

jenkins_pipeline_agent

### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

**Comment by timja** [comment]
> What does this render if the pipeline has `agent none` and there is no `agent` for the stage either? We have some pipelines like that, where an agentless stage sets job properties conditionally before the job starts waiting for an available agent.

Good find, there's a bug it always displays the text when it should only do it if the agent is running on a stage:

image

**Comment by Hammie** [comment]
I've also been testing it on our production instance, and there are some more edge cases where it is not quite working as expected.

- A parallel stage copies the agent property from its first child stage
- A simple single-stage pipeline also shows no agent (even when already running)

I'll try and fix these and also write some tests for them so we can have some confidence in it working well.

But I'm not an experienced Java dev, so we'll see how that goes...

**Comment by Hammie** [comment]
I've added some tests and fixed the known issues.

Not really happy with the actual agent lookup code, but I don't think I can really do it better.

At the very least the tests should make it easier if someone wants to try and rewrite it.

File: src/main/frontend/pipeline-console-view/pipeline-console/main/StageNodeLink.tsx [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import React from "react";
+
+export interface StageNodeLinkProps {
+    agent: string;
+}
+
+const StageNodeLink = ({agent}: StageNodeLinkProps) => {
+    const agentName = agent == "built-in" ? "Jenkins" : agent;
+    const href = `../../../../computer/${agent == "built-in" ? "(built-in)" : agent }/`;
```
Comment (timja) [inline src/main/frontend/pipeline-console-view/pipeline-console/main/StageNodeLink.tsx:9]: I don't think this will be reliable, you could be nested in multi-level folders. you should do it from the root, you can access the relative URL that Jenkins is running on from with (in plain JS, react will be a little different): `document.head.dataset.rooturl`

roughly:
`${document.head.dataset.rooturl}/manage/computer/...`

but you may need a react hook to get the dataset

**Review by timja — COMMENTED** [review]
Code looks pretty good from a read, one improvement needed and I'll do a re-test when I get a chance

File: src/main/frontend/pipeline-console-view/pipeline-console/main/StageNodeLink.tsx (hunk 1 above)
Comment (Hammie) [reply src/main/frontend/pipeline-console-view/pipeline-console/main/StageNodeLink.tsx]: I've updated the code to use the root url. Could not really find a reason why a hook would be needed, so I hope it's good.

File: src/main/frontend/pipeline-console-view/pipeline-console/main/StageNodeLink.tsx [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+function getAgentUrl(name: string) {
+    // Wrap built-in in brackets
+    const id = name == "built-in" ? "(built-in)" : name;
+    const rootPath = document.head.dataset.rootUrl
+    return `${rootPath}/computer/${id}/`;
+}
+
+const StageNodeLink = ({agent}: StageNodeLinkProps) => {
+    const agentName = agent == "built-in" ? "Jenkins" : agent;
```
Comment (timja) [inline src/main/frontend/pipeline-console-view/pipeline-console/main/StageNodeLink.tsx:15]: Doesn't seem to be working the agent name is coming through as `jenkins` and not `Jenkins`:

image

File: src/main/frontend/pipeline-console-view/pipeline-console/main/StageNodeLink.tsx [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import React from "react";
+
+export interface StageNodeLinkProps {
+    agent: string;
+}
+
+function getAgentUrl(name: string) {
+    // Wrap built-in in brackets
+    const id = name == "built-in" ? "(built-in)" : name;
+    const rootPath = document.head.dataset.rootUrl
```
Comment (timja) [inline src/main/frontend/pipeline-console-view/pipeline-console/main/StageNodeLink.tsx:10]: When I run this locally it needs to be all in lowercase:
```suggestion
    const rootPath = document.head.dataset.rooturl
```

**Review by timja — CHANGES_REQUESTED** [review]
The link I am currently getting when I click to view the agent is:
http://localhost:8085/jenkins/job/junit-attachments-test/job/master/31/pipeline-console/undefined/computer/(built-in)/

(with undefined in the url) and the link doesn't work

**Review by timja — APPROVED** [review]
LGTM thanks for your work on this!
